### PR TITLE
Debug symbols improvements

### PR DESCRIPTION
**Pull request by galderz**
This is a draft follow up PR for #9792. Do let me know if I need to open a separate issue for it.

The changes here were partially discussed in [this quarkus-dev discussion](https://groups.google.com/g/quarkus-dev/c/jEJsKmHQzcw/m/XjiVWF9hCgAJ).

Let me address each improvement individually:

### Third party sources

When native debug is enabled, the build step has been enhanced to temporarily copy the `-sources.jar` for each dependency to the ` target/<...>-native-image-source-jar/lib` directory. By doing that, when GraalVM builds the source cache, it'd automatically find the sources for those dependencies and add them to the sources cache. Once native image build completes, these `-sources.jar` in the lib directory are deleted.

In terms of the quarkus-dev discussion above, it uses a solution based on option 1.

There is one detail that still needs to be decided: By default third party sources are not downloaded. Could the native image build step (or any other build step) trigger `mvn dependency:sources` so that those third party sources get downloaded? It'd be nice to achieve that, only when native image debug is enabled? 

The alternative would for the user to invoke `mvn dependency:sources` or the Quarkus sample project generation to add that plugin to the pom.xml of the projects. This is currently required by this draft PR.

### Application sources added to the sources cache

Before this PR, the sources for the user code in the Quarkus application were not included in the sources cache. Fixes in this PR make sure this code also gets added to the sources cache.

### Move the sources cache folder to be under `target/sources-cache`

When using `gdb` with an existing Quarkus application, the existing sources cache folders were very long and not very user friendly to type in gdb. Using one of the debug info options in GraalVM, I've moved the folder to be under `target/sources-cache`.

Note that this change might not make that much sense any more, since @zakkak has improved GraalVM so that `gdb` automatically finds the sources directories, so you no longer have to type those in `gdb` (see https://github.com/oracle/graal/pull/2661).

### Warning

One final warning note. The changes here won't work with any existing official GraalVM releases. If you want to try them out, you'd need to checkout [Mandrel 20.1 branch](https://github.com/graalvm/mandrel/tree/mandrel/20.1) locally, apply fix for https://github.com/oracle/graal/issues/2637, and then use that. Two reasons:

* Adding third party sources that were created 2/3 years before today would trigger https://github.com/oracle/graal/issues/2637. A [PR](https://github.com/oracle/graal/issues/2637) is available but has not yet been integrated.
* The latest official GraalVM 20.1 release does not include all the debug info patches that @adinn created. In particular, that release does not include support for `-H:DebugInfoSourceCacheRoot`

### CODE REVIEW

**Comment by jaikiran** [comment]
Hello @galderz,

> Application sources added to the sources cache
> 
> Before this PR, the sources for the user code in the Quarkus application were not included in the sources cache. Fixes in this PR make sure this code also gets added to the sources cache.

I see that this is being done by adding `src/main/java` to the source search path. I am not sure how the whole native image debugging plays with Kotlin language (which Quarkus supports[1]), but perhaps we should take into account this detail from some build item before deciding whether or not to add `src/main/java` depending on the application's language? Either way, I guess the native-image (at debug or runtime) won't error out if we add this potentionally non-existent directory to its source search path?

[1] https://quarkus.io/guides/kotlin

**Comment by zakkak** [comment]
> Note that this change might not make that much sense any more, since @zakkak has improved GraalVM so that `gdb` automatically finds the sources directories, so you no longer have to type those in `gdb` (see [oracle/graal#2661](https://github.com/oracle/graal/pull/2661)).

IMO It's convenient to have short paths in case manual intervention is needed, so I wouldn't revert that change.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 1]
```diff
@@ -257,6 +268,10 @@ public NativeImageBuildItem build(NativeConfig nativeConfig, NativeImageSourceJa
             }
             if (nativeConfig.debug.enabled) {
                 command.add("-H:GenerateDebugInfo=1");
```
Comment (zakkak) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:270]: Maybe irrelevant to this PR but shouldn't this change to `-g`?

Comment (zakkak) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:270]: NVM `-g` is not available in GraalVM 20.1 CE

**Comment by gsmet** [comment]
@galderz something only targeting Mandrel cannot be included as you're well aware of.

Given we already detect the GraalVM version during the native build here: https://github.com/quarkusio/quarkus/blob/master/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java#L174 , I wonder if we couldn't also detect if it's Mandrel and then apply the additional parameters.

That would allow us to include that PR early.

**Comment by gsmet** [comment]
Just a reminder that 1.7.0.CR1 is planned for Thursday so if you want that in, you'd better hurry :).

**Comment by galderz** [comment]
@gsmet Had PTO. This was a draft PR so wasn't expecting to be integrated as is. I'll close for now and reopen shortly with a final version.

**Pull request closed without merging**